### PR TITLE
Added Rpush, Rpop and RpopCount Commands. (List Commands)

### PR DESCRIPTION
**Pull request by SanHalacogluImproving**

### CODE REVIEW

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+package glide.api.commands;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * List Commands interface.
+ *
+ * @see <a href="https://redis.io/commands/?group=list">List Commands</a>
+ */
+public interface ListCommands {
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:11]: If its going into BaseClient, we should call this `ListBaseCommands`

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    CompletableFuture<Long> rpush(String key, String[] elements);
+
+    /**
+     * Removes and returns the last elements of the list stored at <code>key</code>. The command pops
+     * a single element from the end of the list.
+     *
+     * @see <a href="https://redis.io/commands/rpop/">redis.io</a> for details.
+     * @param key The key of the list.
+     * @return The value of the last element.<br>
+     *     If <code>key</code> does not exist null will be returned.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:33]:
```suggestion
     *     If <code>key</code> does not exist <code>null</code> will be returned.<br>
```

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+     *     If <code>key</code> holds a value that is not a list, an error is raised.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:34]:
```suggestion
     *     If <code>key</code> holds a value that is not a list, an RequestException is thrown.<br>
```
Should we use more specific with error output?

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    CompletableFuture<String> rpop(String key);
+
+    /**
+     * Removes and returns up to <code>count</code> elements from the list stored at <code>key</code>,
+     * depending on the list's length.
+     *
+     * @see <a href="https://redis.io/commands/rpop/">redis.io</a> for details.
+     * @param count The count of the elements to pop from the list.
+     * @returns An array of popped elements will be returned depending on the list's length.<br>
+     *     If <code>key</code> does not exist null will be returned.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:45]:
```suggestion
     *     If <code>key</code> does not exist <code>null</code> will be returned.<br>
```

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+     *     If <code>key</code> holds a value that is not a list, an error is raised.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:46]: same here?

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+     * Insert all the specified values at the tail of the list stored at <code>key</code>. <code>
+     * elements</code> are inserted one after the other to the tail of the list, from the leftmost
+     * element to the rightmost element. If <code>key</code> does not exist, it is created as an empty
+     * list before performing the push operations.
+     *
+     * @see <a href="https://redis.io/commands/rpush/">redis.io</a> for details.
+     * @param key The key of the list.
+     * @param elements The elements to insert at the tail of the list stored at <code>key</code>.
+     * @return The length of the list after the push operations.<br>
+     *     If <code>key</code> holds a value that is not a list, an error is raised.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:22]: Should we mention that an error is thrown?

File: java/client/src/main/java/glide/api/commands/ListCommands.java (hunk 3 above)
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/ListCommands.java:34]: Or better, use the @throws label for this?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 7]
```diff
@@ -169,6 +172,60 @@ public T set(String key, String value, SetOptions options) {
...
+    }
+
+    /**
+     * Removes and returns the last elements of the list stored at <code>key</code>. The command pops
+     * a single element from the end of the list.
+     *
+     * @see <a href="https://redis.io/commands/rpop/">redis.io</a> for details.
+     * @param key The key of the list.
+     * @return Command Response - The value of the last element.<br>
+     *     If <code>key</code> does not exist null will be returned.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:201]:
```suggestion
     *     If <code>key</code> does not exist <code>null</code> will be returned.<br>
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 8]
```diff
@@ -169,6 +172,60 @@ public T set(String key, String value, SetOptions options) {
...
+
+    /**
+     * Removes and returns up to <code>count</code> elements from the list stored at <code>key</code>,
+     * depending on the list's length.
+     *
+     * @see <a href="https://redis.io/commands/rpop/">redis.io</a> for details.
+     * @param count The count of the elements to pop from the list.
+     * @returns Command Response - An array of popped elements will be returned depending on the
+     *     list's length.<br>
+     *     If <code>key</code> does not exist null will be returned.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:219]:
```suggestion
     *     If <code>key</code> does not exist <code>null</code> will be returned.<br>
```

**Review by acarbonetto — APPROVED** [review]
A couple of minor comments

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public interface ListCommands {
+    /**
+     * Insert all the specified values at the tail of the list stored at <code>key</code>. <code>
+     * elements</code> are inserted one after the other to the tail of the list, from the leftmost
+     * element to the rightmost element. If <code>key</code> does not exist, it is created as an empty
+     * list before performing the push operations.
+     *
+     * @see <a href="https://redis.io/commands/rpush/">redis.io</a> for details.
+     * @param key The key of the list.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:19]:
```suggestion
     * @param key The <code>key</code> of the list.
```

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * List Commands interface.
+ *
+ * @see <a href="https://redis.io/commands/?group=list">List Commands</a>
+ */
+public interface ListCommands {
+    /**
+     * Insert all the specified values at the tail of the list stored at <code>key</code>. <code>
+     * elements</code> are inserted one after the other to the tail of the list, from the leftmost
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:14]: Should we replace `list` with `<code>list</code>` everywhere?

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+     * @see <a href="https://redis.io/commands/rpush/">redis.io</a> for details.
+     * @param key The key of the list.
+     * @param elements The elements to insert at the tail of the list stored at <code>key</code>.
+     * @return The length of the list after the push operations.<br>
+     *     If <code>key</code> holds a value that is not a list, an error is raised.<br>
+     */
+    CompletableFuture<Long> rpush(String key, String[] elements);
+
+    /**
+     * Removes and returns the last elements of the list stored at <code>key</code>. The command pops
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:27]:
```suggestion
     * Remove and return the last elements of the list stored at <code>key</code>. The command pops
```

File: java/client/src/main/java/glide/api/commands/ListCommands.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/ListCommands.java:34]: Should be everywhere then (for almost all commands)
A subject for another PR. Do you want to create a task for this @acarbonetto ?

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+     * @see <a href="https://redis.io/commands/rpop/">redis.io</a> for details.
+     * @param key The key of the list.
+     * @return The value of the last element.<br>
+     *     If <code>key</code> does not exist null will be returned.<br>
+     *     If <code>key</code> holds a value that is not a list, an error is raised.<br>
+     */
+    CompletableFuture<String> rpop(String key);
+
+    /**
+     * Removes and returns up to <code>count</code> elements from the list stored at <code>key</code>,
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:39]:
```suggestion
     * Remove and return up to <code>count</code> elements from the list stored at <code>key</code>,
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 13]
```diff
@@ -169,6 +172,60 @@ public T set(String key, String value, SetOptions options) {
...
+     */
+    public T rpush(@NonNull String key, @NonNull String[] elements) {
+        ArgsArray commandArgs = buildArgs(ArrayUtils.addFirst(elements, key));
+
+        protobufTransaction.addCommands(buildCommand(RPush, commandArgs));
+        return getThis();
+    }
+
+    /**
+     * Removes and returns the last elements of the list stored at <code>key</code>. The command pops
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:195]:
```suggestion
     * Remove and return the last elements of the list stored at <code>key</code>. The command pops
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 14]
```diff
@@ -169,6 +172,60 @@ public T set(String key, String value, SetOptions options) {
...
     }
 
+    /**
+     * Insert all the specified values at the tail of the list stored at <code>key</code>. <code>
+     * elements</code> are inserted one after the other to the tail of the list, from the leftmost
+     * element to the rightmost element. If <code>key</code> does not exist, it is created as an empty
+     * list before performing the push operations.
+     *
+     * @see <a href="https://redis.io/commands/rpush/">redis.io</a> for details.
+     * @param key The key of the list.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:182]:
```suggestion
     * @param key The <code>key</code> of the list.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 15]
```diff
@@ -169,6 +172,60 @@ public T set(String key, String value, SetOptions options) {
...
+     */
+    public T rpop(@NonNull String key) {
+        ArgsArray commandArgs = buildArgs(key);
+
+        protobufTransaction.addCommands(buildCommand(RPop, commandArgs));
+        return getThis();
+    }
+
+    /**
+     * Removes and returns up to <code>count</code> elements from the list stored at <code>key</code>,
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:212]:
```suggestion
     * Remove and return up to <code>count</code> elements from the list stored at <code>key</code>,
```

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+                .toArray(String[]::new);
+    }
+
+    /**
+     * Casts an array of objects to an array of type T.
+     *
+     * @param objectArr Array of objects to cast.
+     * @param clazz The class of the array elements to cast to.
+     * @return An array of type T, containing the elements from the input array.
+     * @param <T> The type to which the elements are cast.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/utils/ArrayTransformUtils.java:29]: please add `@param <U>` to javadoc

File: java/client/src/main/java/glide/utils/ArrayTransformUtils.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.utils;
+
+import java.lang.reflect.Array;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.stream.Stream;
+
+public class ArrayTransformUtils {
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/utils/ArrayTransformUtils.java:9]: Please add a javadoc for the class

File: java/client/src/main/java/glide/api/commands/ListCommands.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.api.commands;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * List Commands interface.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/ListCommands.java:7]:
```suggestion
 * List Commands interface for both standalone and cluster clients.
```

**Review by Yury-Fridlyand — APPROVED** [review]
Minor javadoc updates are needed

**Pull request closed without merging**